addons/website_sale_options/models/sale_order.py

```diff
@@ -33,26 +33,30 @@ def _cart_find_product_line(self, cr, uid, ids, product_id=None, line_id=None, c
     def _cart_update(self, cr, uid, ids, product_id=None, line_id=None, add_qty=0, set_qty=0, context=None, **kwargs):
         """ Add or set product quantity, add_qty can be negative """
         value = super(sale_order, self)._cart_update(cr, uid, ids, product_id, line_id, add_qty, set_qty, context=context, **kwargs)
-        
-        linked_line_id = kwargs.get('linked_line_id')
+
         sol = self.pool.get('sale.order.line')
         line = sol.browse(cr, SUPERUSER_ID, value.get('line_id'), context=context)
 
-        for so in self.browse(cr, uid, ids, context=context):
-
-            if linked_line_id and linked_line_id in map(int,so.order_line):
-                linked = sol.browse(cr, SUPERUSER_ID, linked_line_id, context=context)
-                line.write({
-                        "name": _("%s\nOption for: %s") % (line.name, linked.product_id.name_get()[0][1]),
-                        "linked_line_id": linked_line_id
-                    })
+        # link a product to the sale order
+        if kwargs.get('linked_line_id'):
+            linked_line_id = sol.browse(cr, SUPERUSER_ID, kwargs.get('linked_line_id'), context=context)
+            line.write({
+                    "name": _("%s\nOption for: %s") % (line.name, linked_line_id.product_id.name_get()[0][1]),
+                    "linked_line_id": linked_line_id.id
+                })
 
+        value['option_ids'] = set()
+        for so in self.browse(cr, uid, ids, context=context):
             # select linked product
-            option_ids = [l for l in so.order_line if l.linked_line_id.id == line.id]
+            option_line_ids = [l for l in so.order_line if l.linked_line_id.id == line.id]
 
             # update line
-            for l in option_ids:
-                super(sale_order, self)._cart_update(cr, uid, ids, l.product_id.id, l.id, add_qty, set_qty, context=context, **kwargs)
+            for option_line_id in option_line_ids:
+                super(sale_order, self)._cart_update(cr, uid, ids, option_line_id.product_id.id, option_line_id.id, add_qty, set_qty, context=context, **kwargs)
+                option_line_id.write({"name": _("%s\nOption for: %s") % (option_line_id.name, option_line_id.linked_line_id.product_id.name_get()[0][1])})
+
+            value['option_ids'] |= set(map(int, option_line_ids))
```

Comment:
This one is just unbearable. can't you just do one of the two below:
```
value['option_ids'].union(option_line_ids.ids)
``` 
or, within the for loop just above
```
value['option_ids'].add(option_line_id.id)
``` 
Your code must always be the most readable possible for a newbie. Always thing that someone will have to read your code after you wrote it.